segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/TableBasedMetadataStore.java

```diff
@@ -0,0 +1,186 @@
+    @Override
+    protected TransactionData read(String key) throws StorageMetadataException {
+        ensureInitialized();
+        List<BufferView> keys = new ArrayList<>();
+        keys.add(new ByteArraySegment(key.getBytes()));
+        try {
+            List<TableEntry> retValue = this.tableStore.get(tableName, keys, timeout).get();
+            if (retValue.size() == 1) {
+                TableEntry entry = retValue.get(0);
+                if (null != entry) {
+                    val arr = entry.getValue();
+                    TransactionData txnData = serializer.deserialize(arr);
+                    txnData.setDbObject(entry.getKey().getVersion());
+                    txnData.setPersisted(true);
+                    return txnData;
+                }
+            }
+        } catch (IllegalStateException e) {
+            throw e;
+        } catch (Exception e) {
+            throw new StorageMetadataException("Error while reading", e);
+        }
+
+        return TransactionData.builder()
+                .key(key)
+                .persisted(true)
+                .dbObject(TableKey.NOT_EXISTS)
+                .build();
+    }
+
+    /**
+     * Writes transaction data from a given list to the metadata store.
+     *
+     * @param dataList List of transaction data to write.
+     * @throws StorageMetadataException Exception related to storage metadata operations.
+     */
+    @Override
+    protected void writeAll(Collection<TransactionData> dataList) throws StorageMetadataException {
+        ensureInitialized();
+        List<TableEntry> toUpdate = new ArrayList<>();
+        HashMap<TableEntry, TransactionData> entryToTxnDataMap = new HashMap<TableEntry, TransactionData>();
+        HashMap<TableKey, TransactionData> deletedKeyToTxnDataMap = new HashMap<TableKey, TransactionData>();
+        List<TableKey> keysToDelete = new ArrayList<>();
+        try {
+            for (TransactionData txnData : dataList) {
+                Preconditions.checkState(null != txnData.getDbObject());
+
+                long version = ((Long) txnData.getDbObject()).longValue();
+                if (null == txnData.getValue()) {
+                    val toDelete = TableKey.versioned(new ByteArraySegment(txnData.getKey().getBytes()),
+                            TableKey.NO_VERSION);
+                    keysToDelete.add(toDelete);
+                    deletedKeyToTxnDataMap.put(toDelete, txnData);
+                }
+
+                val arraySegment = serializer.serialize(txnData);
+
+                TableEntry tableEntry = TableEntry.versioned(
+                        new ByteArraySegment(txnData.getKey().getBytes()),
+                        arraySegment,
+                        version);
+                entryToTxnDataMap.put(tableEntry, txnData);
+                toUpdate.add(tableEntry);
+            }
+
+            // Now put uploaded keys.
+            List<Long> ret = this.tableStore.put(tableName, toUpdate, timeout).get();
```

Comment:
Do you really need to do this? This will block this thread and use another thread to execute that call. It is a recipe for running out of threads very very quickly.

Please rewrite this method to be a CompletableFuture (async) so it doesn't have this problem.

Reply:
Given the initial uncertainty with the approach and added complexity I decided to do inital implementation without async apis. Once things are solid we can very much turn everything into async relatively easily.  The point - the table segment access is already in a synchronized block to maintain consistency guarantees. So that waiting on future is not causing any additional stalls. In near future, I do absolutely plan to remove that big synchronization point and also turn this whole thing in async.

Reply:
I'll create an issue to track this.

Reply:
Currently table segment can not have update and delete as part of same transaction. (you can't add/update a few keys and also delete a few as part of same transaction.)
That is the reason why I have to delete the keys separately.
These two could be executed in parallel.

Most updates are with either 1 or 2 keys.

Reply:
Delete is done in two steps .
1. Update key with value corresponding to null . 
2. Delete the key 
All keys in a single transaction are updated simultaneously. Update could fail because of any number of reason - eg. Version mismatch one of the (non-deleted) keys.

In all cases, the delete should not succeed if update fails earlier. Therefore it has to wait on update.